libs/knowledge-store/ragstack_knowledge_store/knowledge_store.py

```diff
@@ -46,6 +49,39 @@ def _results_to_ids(results: Optional[ResponseFuture]) -> Iterable[str]:
         for row in results:
             yield row.content_id
 
+def emb_to_ndarray(embedding: List[float]) -> np.ndarray:
+    embedding = np.array(embedding, dtype=np.float32)
+    if embedding.ndim == 1:
+        embedding = np.expand_dims(embedding, axis=0)
+    return embedding
+
+@dataclass
+class _Candidate:
+  score: float
+  similarity_to_query: float
+  """Lambda * Similarity to the question."""
+
+  embedding: np.ndarray
+  """Embedding used for updating similarity to selections."""
+
+  redundancy: float
+  """(1 - Lambda) * max(Similarity to selected items)."""
+
+  def __init__(self, embedding: List[float], lambda_mult: float, query_embedding: np.ndarray):
+    self.embedding = emb_to_ndarray(embedding)
+
+    # TODO: Refactor to use cosine_similarity_top_k to allow an array of embeddings?
```

Comment:
I need to rework somethings to make this take advantage of the fact that cosine_similarity operates on a matrix (I believe) so that I can do all of the comparisons in one step. But, this works for now. Should that be deferred, or is it worth doing in this iteration?

Reply:
How about using `langchain_community.utils.math.cosine_similarity` ? It uses numpy and has optional support for simd.

Reply:
Oh, right. I read too quickly.